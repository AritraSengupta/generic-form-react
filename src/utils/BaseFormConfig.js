export default class BaseFormConfig {
  constructor(props) {
    this.props = props
  }

  title = 'Generic Conditional Input Form'

  data = () => []

  reSyncValues = async (values, prevValues, errors) => ({ values, errors })
}
