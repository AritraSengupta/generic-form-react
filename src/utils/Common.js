export function changeToUSDateStyle(value) {
  const arr = value.split('-')
  const usStyle = [arr[1], arr[0], arr[2]].join('-')
  return usStyle
}

export function isValidDate(d) {
  return d instanceof Date && !isNaN(d)
}

export function formatInputIfDate(date) {
  if (!date || typeof date !== 'string') return date
  const dateArray = date.split('-')
  if (dateArray.length === 3 && !isValidDate(new Date(date))) {
    // Try to format if not a valid date
    const openApiDate = `${dateArray[2]}-${dateArray[1]}-${dateArray[0]}`
    const dateObj = new Date(openApiDate)
    if (isValidDate(dateObj)) {
      return openApiDate
    }
  }
  return date
}

export function getDateString(date, month, year) {
  return `${date}-${month + 1}-${year}`
}

export const getFormattedDate = (date) => {
  if (!isValidDate(new Date(date))) return date

  const dateObj = new Date(date)
  return getDateString(
    dateObj.getDate(),
    dateObj.getMonth(),
    dateObj.getFullYear()
  )
}

export const formatToFixed = (value, decimalPlaces, defaultValue) => {
  const numericalValue = parseFloat(value).toFixed(decimalPlaces)
  if (!isNaN(numericalValue)) {
    return numericalValue
  }
  return defaultValue
}

export const twoDecimals = (value, decimalPoints) => {
  if (value === 0) return 0
  const t = value ? `${value}` : ''
  if (Number(t) < 0 || t === '-') {
    return ''
  }
  const newValue =
    t.indexOf('.') >= 0
      ? t.substr(0, t.indexOf('.')) +
        t.substr(t.indexOf('.'), decimalPoints + 1)
      : t
  return newValue
}

export const convertToArray = (hashMap) =>
  Object.keys(hashMap).filter((key) => hashMap[key])
