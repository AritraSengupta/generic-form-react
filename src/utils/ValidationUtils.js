/**
 * This file contains all the utils regarding the form validation process,
 * We will use the following structure for error:
 * @param error: {
 *  name: <string> 'field name',
 *  state: <boolean> 'true or false',
 *  message: <string> message to be shown to the user in case state is true
 * }
 */

export const standardValidators = {
  required: (value) => {
    if (typeof value === 'number') {
      return {
        state: false
      }
    }
    return {
      state: !(value && value.length),
      message: 'is required'
    }
  },

  maxLength: (value, { maxLength }) => {
    if (typeof value !== 'string') {
      return {
        state: true,
        message: 'should be a string'
      }
    }
    return {
      state: value.length > maxLength,
      message: `can be a maximum length of ${maxLength}`
    }
  },

  minLength: (value, { minLength }) => {
    if (typeof value !== 'string') {
      return {
        state: true,
        message: 'should be a string'
      }
    }
    return {
      state: value.length < minLength,
      message: `should be a minimum length of ${minLength}`
    }
  },

  number: (value) => ({
    state: !(typeof value === 'number' && !Number.isNaN(Number(value))),
    message: 'should be a number'
  }),

  max: (value, { max }) => ({
    state: Number(value) > max,
    message: `cannot be more than ${max}`
  }),

  min: (value, { min }) => ({
    state: Number(value) < min,
    message: `cannot be less than ${min}`
  }),

  pattern: (value, { pattern, message }) => {
    const regPattern = new RegExp(pattern)
    return {
      state: !regPattern.test(value),
      message: message || `does not match pattern ${pattern}`
    }
  }
}

export const validateValue = (value, validator) => {
  if (typeof validator === 'function') {
    return validator(value)
  }

  if (validator.type && standardValidators[validator.type]) {
    return standardValidators[validator.type](value, validator.params)
  }

  throw new Error('Not a valid Validator')
}

export const runValidationAgainstAllValues = (values, validator) => {
  if (Array.isArray(values) && values.length) {
    let error = null
    for (let i = 0; i < values.length; i += 1) {
      if (error && error.state) break
      error = validateValue(values[i], validator)
    }
    return error
  }
  return validateValue(values, validator)
}

export const runValidation = (values, validators, name) => {
  if (Array.isArray(validators)) {
    const errors = []
    validators.forEach((validator) => {
      const error = runValidationAgainstAllValues(values, validator)

      if (error.state) {
        errors.push({ name, ...error })
      }
    })
    return errors
  }
  const error = runValidationAgainstAllValues(values, validators)
  return error.state ? [{ name, ...error }] : []
}

export const isRequired = (validators) => {
  if (Array.isArray(validators)) {
    return validators.some((v) => !!(v && v.type === 'required'))
  }
  return !!(validators && validators.type === 'required')
}
