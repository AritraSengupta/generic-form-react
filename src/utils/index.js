export * from './ValidationUtils'
export * from './Common'
export * from './FormUtils'
export { default as BaseFormConfig } from './BaseFormConfig'
