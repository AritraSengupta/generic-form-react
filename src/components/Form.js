import React from 'react'
import { isEqual } from 'lodash'

import FormDisplay from './FormDisplay'
import { getDefaultValues } from '../utils'

/**
 * GenericConditionalForm component renders a form which can be conditional on the current data
 * @props (onChange): OPTIONAL: Function called whenever a formState changes
 * @props (form): the form whose format is very specfic: FormType
 * @props (defaultValues): OPTIONAL: To set the initial values of the form
 * @props (errors): OPTIONAL: To set the initial errors of the form
 * type FormType = (defaultValues: DefaultValues) => {
 *    title: String
 *    data: Array<FormUnit>
 * }
 *
 * type DefaultValues = {[ dataId: String]: { defaultValue: String | { value: Array | String, selected: Array<Textvaluepair> | TextValuePair }, options?: Array<TextValuePair> }}
 * type FormUnit = {
 *    fieldname: String,
 *    dataId: String,
 *    type: 'input' | 'checkbox' | 'dropdown' | 'radio' | 'files' | 'hidden' | 'textarea' | 'date' | 'paymentTerms',
 *    config: Object<any>,
 *    defaultValue: DefaultValue,
 *    validation?: String,
 *    options?: Array<TextValuePair>,
 * }
 * Send the form as a function which only accepts defaultValue object and fill out the rest
 */

export class Form extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      defaultValues: props.defaultValues || {},
      prevValues: {},
      errors: props.errors || {},
      loading: false
    }
    this.formRef = React.createRef()
    this.onChange = this.onChange.bind(this)
    this.updateForm = this.updateForm.bind(this)
    this.validateForm = this.validateForm.bind(this)
  }

  async onChange({ formData, errors }) {
    const {
      defaultValues,
      formData: oldFormData,
      errors: oldErrors
    } = this.state
    if (isEqual(oldFormData, formData) && isEqual(errors, oldErrors)) return
    const { data: form, reSyncValues } = this.props
    const prevForm = form(defaultValues)
    let newDefaultValues = getDefaultValues(prevForm, formData)
    let finalErrors = errors
    let loading = false
    if (reSyncValues) {
      try {
        loading = true
        this.updateForm(
          newDefaultValues,
          defaultValues,
          finalErrors,
          formData,
          loading
        )
        const { values, errors: newErrors } = await reSyncValues(
          newDefaultValues,
          defaultValues,
          errors,
          formData,
          this.updateForm
        )
        loading = false
        newDefaultValues = values
        finalErrors = newErrors
      } catch (e) {
        loading = false
        finalErrors.generic = true
      }
    }

    this.updateForm(
      newDefaultValues,
      defaultValues,
      finalErrors,
      formData,
      loading
    )
  }

  validateForm() {
    return this.formRef.current.validateForm()
  }

  updateForm(defaultValues, prevValues, errors, formData, loading) {
    const { onChange } = this.props

    this.setState({
      defaultValues,
      prevValues,
      errors,
      formData,
      loading
    })
    if (onChange) {
      onChange({ formData, errors })
    }
  }

  render() {
    const { errors, defaultValues, prevValues, loading } = this.state
    const { data: form, columns } = this.props
    const currentForm = form(defaultValues, prevValues)
    return (
      <FormDisplay
        columns={columns}
        data={currentForm}
        onChange={this.onChange}
        errors={errors}
        loading={loading}
        ref={this.formRef}
      />
    )
  }
}
