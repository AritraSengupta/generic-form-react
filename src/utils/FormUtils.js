import { get, set, uniqBy, has } from 'lodash'

import { formatInputIfDate } from './Common'

/**
 * @param {*} formData -> the form datastructure object
 * @param {*} formState -> the current state of the form
 * This function returns the default values of the current form in the format which
 * the form understands. Should be used when converting formState to defaultvalue object
 * to be passed to the form
 */

export const getDefaultValues = (formData, formState) => {
  const defaultValues = {}
  formData.forEach((data) => {
    if (data.type === 'checkbox') {
      const defaultValue = get(formState, data.dataId, {
        value: {},
        selected: []
      })
      defaultValues[data.dataId] = {
        defaultValue
      }
      return
    }
    if (data.type === 'files') {
      const defaultValue = get(formState, data.dataId, [])
      defaultValues[data.dataId] = { defaultValue }
      return
    }
    if (data.type === 'dropdown') {
      const defaultValue = get(formState, data.dataId, {})
      let value
      let options
      if (data.config && data.config.multiple) {
        value = (defaultValue && defaultValue.selected) || []
        if (value.length === 0) {
          defaultValue.value = []
          defaultValue.selected = []
        }
        options = uniqBy([...(data.options || []), ...value], 'value')
      } else {
        const { selected } = defaultValue || {}
        value = selected ? [selected] : []
        options = uniqBy([...(data.options || []), ...value], 'value')
      }

      defaultValues[data.dataId] = {
        defaultValue,
        options
      }
      return
    }
    const defaultValue = get(formState, data.dataId, null)
    defaultValues[data.dataId] = {
      defaultValue
    }
  })
  return defaultValues
}

/**
 *
 * @param {*} formDataFunc: The conditional Form Function
 * @param {*} apiData: apiData of that needs mapped
 * @param {*} dataMap: the datamap to map apidata to formdata
 *
 * returns an object with keys as form ids and value with default values from api
 * NOTE options from dropdown are not mapped
 */
export const conditionalFormDefaultValuesGetter = (
  formDataFunc,
  apiData,
  dataMap
) => {
  const formData = formDataFunc({})
  const mappedDefaults = {}
  formData.forEach((data) => {
    if (data.type === 'checkbox') {
      const defaultValue = get(apiData, dataMap[data.dataId], null)
      if (defaultValue === null || defaultValue === undefined) {
        // case to handle null values for api calls
        mappedDefaults[data.dataId] = {
          defaultValue: null
        }
      } else {
        const value = (defaultValue || []).reduce((acc, curr) => {
          acc[curr] = true
          return acc
        }, {})
        mappedDefaults[data.dataId] = {
          defaultValue: {
            value,
            selected: defaultValue
          }
        }
      }
    } else if (data.type === 'files') {
      const defaultValue = get(apiData, dataMap[data.dataId], [])
      mappedDefaults[data.dataId] = {
        defaultValue: {
          files: defaultValue,
          deletedFiles: []
        }
      }
    } else if (data.type === 'dropdown') {
      const defaultValue = get(apiData, dataMap[data.dataId].id, null)
      if (defaultValue === null || defaultValue === undefined) {
        // case to handle null api values for api calls
        mappedDefaults[data.dataId] = {
          defaultValue: null,
          options: data.options
        }
      } else {
        const valueLabel = dataMap[data.dataId].value || 'id'
        const textLabel = dataMap[data.dataId].text || 'name'
        let value
        let options
        if (data.config && data.config.multiple) {
          const selected =
            (defaultValue &&
              defaultValue.map((d) => ({
                value: d[valueLabel],
                text: d[textLabel]
              }))) ||
            []
          value = {
            value:
              (defaultValue && defaultValue.map((d) => d[valueLabel])) || [],
            selected
          }
          options = uniqBy([...data.options, ...selected], 'value')
        } else {
          const selected = {
            text: defaultValue && defaultValue[textLabel],
            value: defaultValue && defaultValue[valueLabel]
          }
          value = { value: defaultValue && defaultValue[valueLabel], selected }
          options = uniqBy([...data.options, selected], 'value')
        }
        mappedDefaults[data.dataId] = {
          defaultValue: {
            ...value
          },
          options
        }
      }
    } else {
      const defaultValue = get(apiData, dataMap[data.dataId], null)
      mappedDefaults[data.dataId] = {
        defaultValue
      }
    }
  })
  return mappedDefaults
}

export const defaultValueMapper = (formData, apiData, dataMap) =>
  formData.map((data) => {
    if (data.type === 'checkbox') {
      const defaultValue = get(apiData, dataMap[data.dataId], []) || []
      const value = defaultValue.reduce((acc, curr) => {
        acc[curr] = true
        return acc
      }, {})
      return {
        ...data,
        defaultValue: { value, selected: defaultValue }
      }
    }
    if (data.type === 'files') {
      const defaultValue = get(apiData, dataMap[data.dataId], [])
      return {
        ...data,
        defaultValue: { files: defaultValue, deletedFiles: [] }
      }
    }
    if (data.type === 'dropdown') {
      const defaultValue = get(apiData, dataMap[data.dataId].id, null)
      const valueLabel = dataMap[data.dataId].value || 'id'
      const textLabel = dataMap[data.dataId].text || 'name'
      let value
      let options
      if (data.config && data.config.multiple) {
        const selected =
          (defaultValue &&
            defaultValue.map((d) => ({
              value: d[valueLabel],
              text: d[textLabel]
            }))) ||
          []
        value = {
          value: (defaultValue && defaultValue.map((d) => d[valueLabel])) || [],
          selected
        }
        options = uniqBy([...(data.options || []), ...selected], 'value')
      } else {
        const selected = {
          text: defaultValue && defaultValue[textLabel],
          value: defaultValue && defaultValue[valueLabel]
        }
        value = { value: defaultValue && defaultValue[valueLabel], selected }
        options = uniqBy([...(data.options || []), selected], 'value')
      }
      return {
        ...data,
        defaultValue: value,
        options
      }
    }
    const defaultValue = get(apiData, dataMap[data.dataId], null)
    return {
      ...data,
      defaultValue
    }
  })

export const apiDataMapper = (formData, dataMap) => {
  const apiData = {}
  Object.keys(formData).forEach((key) => {
    let formDataValue
    // Doing it this way ensures that is the selected key is present then the value
    // inside selected is retured regardless whether its falsy or not
    if (has(formData, `${key}.selected`)) {
      formDataValue = formData[key].selected
    } else if (has(formData, `${key}.files`)) {
      formDataValue = formData[key].files
    } else {
      // Change to OpenAPI spec if the date is in DD-MM-YYYY format
      formDataValue = formatInputIfDate(formData[key])
    }
    const formKey = dataMap[key].id || dataMap[key]
    set(apiData, formKey, formDataValue)
  })
  return apiData
}

export const dropdownOptionsMapper = (
  list,
  valueLabel = 'id',
  textLabel = 'name'
) => {
  if (!list) return []
  return list.map((l) => ({
    text: l[textLabel],
    value: l[valueLabel]
  }))
}

export const getValidatorMap = (form, formState) =>
  form.map((el) => {
    let value = formState[el.dataId]
    if (value && el.type === 'dropdown') {
      const { selected } = value
      if (Array.isArray(selected)) {
        value = selected.map((s) => s.text)
      } else {
        value = selected && selected.text
      }
    } else if (value && el.type === 'checkbox') {
      value = value.selected
    }
    return {
      id: el.dataId,
      validators: el.validators,
      value,
      name: el.fieldname
    }
  })

export const getFormGridStruct = (formData, columns) => {
  const gridList = []
  let i = 0
  while (i < formData.length) {
    const gridRow = []
    let j = 0
    while (j < columns && i < formData.length) {
      if (formData[i] && formData[i].type !== 'hidden') {
        gridRow.push(formData[i])
        j += 1
      }
      i += 1
    }

    if (gridRow.length) {
      gridList.push(gridRow)
    }
  }
  return gridList
}

export const getErrorList = (errors) => {
  const errorList = []
  Object.keys(errors).forEach((key) => {
    const valErrors = []
    ;(errors[key] || []).forEach((e) => {
      e.state && valErrors.push(`${e.name}: ${e.message}`)
    })
    errorList.push(...valErrors)
  })
  return errorList
}

export function getDefaultState(props) {
  return {
    formData: props.data.reduce((acc, curr) => {
      acc[curr.dataId] = curr.defaultValue
      return acc
    }, {}),
    errors: props.errors || {},
    dynamicOptions: props.data.reduce((acc, curr) => {
      if (curr.type === 'dropdown' && curr.config && curr.config.search) {
        acc[curr.dataId] = {
          options: curr.options,
          callback: curr.config.callback,
          loading: false
        }
      }
      return acc
    }, {})
  }
}
