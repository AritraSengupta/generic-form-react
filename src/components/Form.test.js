import React from 'react'

import { Form } from './Form'
import { BaseFormConfig } from '../utils'

const searchFunction = async () =>
  new Promise((resolve) =>
    setTimeout(
      () =>
        resolve([
          { text: 'Option1', value: 1 },
          { text: 'Option2', value: 2 },
          { text: 'Option3', value: 3 },
          { text: 'Option4', value: 4 }
        ]),
      1000
    )
  )

class ConditionalFormAdvanced extends BaseFormConfig {
  title = 'Basic Example'

  data = (currValue) => {
    return [
      {
        dataId: 'id',
        type: 'hidden',
        defaultValue: null
      },
      {
        fieldname: 'Date',
        dataId: 'date',
        type: 'date',
        defaultValue: currValue.date && currValue.date.defaultValue,
        validators: [{ type: 'required' }]
      },
      {
        fieldname: 'Name',
        dataId: 'name',
        type: 'input',
        defaultValue: currValue.name && currValue.name.defaultValue,
        validators: [{ type: 'required' }]
      },
      {
        fieldname: 'TextArea',
        dataId: 'textarea',
        type: 'textarea',
        defaultValue: currValue.textarea && currValue.textarea.defaultValue,
        validators: [{ type: 'required' }]
      },
      {
        fieldname: 'Dropdown',
        dataId: 'dropdown',
        type: 'dropdown',
        defaultValue: currValue.dropdown && currValue.dropdown.defaultValue,
        config: {
          search: true,
          callback: searchFunction,
          clearable: true
        },
        options: (currValue.dropdown && currValue.dropdown.options) || [],
        validators: [{ type: 'required' }]
      },
      {
        fieldname: 'Dropdown Multiple',
        dataId: 'dropdownMul',
        type: 'dropdown',
        defaultValue: (currValue.dropdown &&
          currValue.dropdownMul.defaultValue) || { value: [], selected: [] },
        config: {
          multiple: true,
          clearable: true
        },
        options: (currValue.dropdown && currValue.dropdownMul.options) || [],
        validators: [{ type: 'required' }]
      },
      {
        fieldname: 'Checkbox',
        dataId: 'checkbox',
        type: 'checkbox',
        defaultValue: currValue.checkbox && currValue.checkbox.defaultValue,
        options: [
          { text: 'Checkbox1', value: 1 },
          { text: 'Checkbox2', value: 2 }
        ],
        validators: [{ type: 'required' }]
      },
      {
        fieldname: 'Radio',
        dataId: 'radio',
        type: 'radio',
        defaultValue: currValue.radio && currValue.radio.defaultValue,
        options: [
          { text: 'Radio1', value: 1 },
          { text: 'Radio2', value: 2 },
          { text: 'Radio3', value: 3 }
        ]
      }
    ]
  }

  reSyncValues = async (currValues, prevValues, errors) => {
    const values = { ...currValues }
    const currDropdownValue =
      currValues.dropdown &&
      currValues.dropdown.defaultValue &&
      currValues.dropdown.defaultValue.value
    const prevDropdownValue =
      prevValues.dropdown &&
      prevValues.dropdown.defaultValue &&
      prevValues.dropdown.defaultValue.value

    if (currDropdownValue !== prevDropdownValue) {
      const options = await searchFunction()
      const firstChoice = options[0]
      values.dropdownMul = {
        defaultValue: { value: firstChoice.value, selected: firstChoice },
        options
      }
    }

    return {
      values,
      errors
    }
  }
}

describe('Form', () => {
  it('should render the form', () => {
    const conditionalFormAdvanced = new ConditionalFormAdvanced()
    const onChangeMock = jest.fn()
    const { container } = render(
      <Form
        data={conditionalFormAdvanced.data}
        reSyncValues={conditionalFormAdvanced.reSyncValues}
        onChange={onChangeMock}
      />
    )

    expect(container).toMatchSnapshot()
    expect(onChangeMock).toHaveBeenCalled()
  })
})
