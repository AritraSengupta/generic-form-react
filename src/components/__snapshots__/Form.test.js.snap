// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render the form 1`] = `
<div>
  <form
    class="ui loading form"
  >
    <div
      class="ui error message"
    >
      <div
        class="content"
      >
        <div
          class="header"
        >
          Please check the form for the following
        </div>
        <ul
          class="list"
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Date
      </label>
      <div
        class="ui fluid left icon input"
      >
        <input
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          class="calendar icon"
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Name
      </label>
      <div
        class="ui input"
        data-testid="input"
        field_name="Name"
        validators="[object Object]"
      >
        <input
          name="name"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        TextArea
      </label>
      <textarea
        fieldname="TextArea"
        name="textarea"
        required=""
        rows="3"
        validators="[object Object]"
      />
    </div>
    <div
      class="required field"
    >
      <label>
        Dropdown
      </label>
      <div
        aria-busy="false"
        aria-disabled="false"
        aria-expanded="false"
        class="ui search selection dropdown"
        data-testid="dropdown-search"
        fieldname="Dropdown"
        name="dropdown"
        required=""
        role="combobox"
        validators="[object Object]"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          class="search"
          tabindex="0"
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
          role="listbox"
        >
          <div
            class="message"
          >
            No results found.
          </div>
        </div>
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Dropdown Multiple
      </label>
      <div
        aria-expanded="false"
        aria-multiselectable="true"
        class="ui multiple selection dropdown"
        data-testid="dropdown"
        fieldname="Dropdown Multiple"
        name="dropdownMul"
        required=""
        role="listbox"
        tabindex="0"
        validators="[object Object]"
      >
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Checkbox
      </label>
    </div>
    <div
      class="field"
    >
      <div
        class="ui checkbox"
        fieldname="Checkbox"
        validators="[object Object]"
      >
        <input
          class="hidden"
          name="checkbox"
          readonly=""
          tabindex="0"
          type="checkbox"
          value="1"
        />
        <label>
          Checkbox1
        </label>
      </div>
    </div>
    <div
      class="field"
    >
      <div
        class="ui checkbox"
        fieldname="Checkbox"
        validators="[object Object]"
      >
        <input
          class="hidden"
          name="checkbox"
          readonly=""
          tabindex="0"
          type="checkbox"
          value="2"
        />
        <label>
          Checkbox2
        </label>
      </div>
    </div>
    <div
      class="field"
    >
      <label>
        Radio
      </label>
    </div>
    <div
      class="field"
    >
      <div
        class="ui radio checkbox"
        fieldname="Radio"
      >
        <input
          class="hidden"
          name="radio"
          readonly=""
          tabindex="0"
          type="radio"
          value="1"
        />
        <label>
          Radio1
        </label>
      </div>
    </div>
    <div
      class="field"
    >
      <div
        class="ui radio checkbox"
        fieldname="Radio"
      >
        <input
          class="hidden"
          name="radio"
          readonly=""
          tabindex="0"
          type="radio"
          value="2"
        />
        <label>
          Radio2
        </label>
      </div>
    </div>
    <div
      class="field"
    >
      <div
        class="ui radio checkbox"
        fieldname="Radio"
      >
        <input
          class="hidden"
          name="radio"
          readonly=""
          tabindex="0"
          type="radio"
          value="3"
        />
        <label>
          Radio3
        </label>
      </div>
    </div>
  </form>
</div>
`;
